### PR TITLE
Refactor layer construction code for Maven and Gradle

### LINKED ISSUE #1065: Try alternative mini framework for constructing Java layers
From https://github.com/GoogleContainerTools/jib/pull/1062#pullrequestreview-159975210

### LINKED ISSUE #1089: Reuse code for JavaLayerConfigurations for WAR project between Maven and Gradle
The logic for configuring WAR layers should be identical between Maven and Gradle and easy to share code, since you always work on an exploded WAR directory and create layers in the same way.

### LINKED ISSUE #1115: Add non-recursive addEntry methods to JavaLayerConfigurations
Useful for the case explained here: https://github.com/GoogleContainerTools/jib/pull/1049#discussion_r221093432

We are actually adding files non-cursively since we are walking a full directory hierarchy in `GradleLayerConfigurations` and `MavenlayerConfigurations`, and without the non-recursive version, we have to check if empty directories are empty so that it's safe to pass the empty directories to the recursive version.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #1089.
Fixes #1115.
Closes #1065.

- Moved the WAR layer construction code to `JavaLayerConfigurationsHelper.fromExplodedWar` in `jib-plugins-common` to share it between Maven and Gradle.
- Pushed down to `JavaLayerConfigurations` the helper method in Maven that recursively adds files to a layer after filtering. Simplied layer construction in Maven and Gradle using it.

### CODE REVIEW

**Comment by briandealwis** [comment]
To my mind, `JavaLayerConfiguration` should really be treated as a configurator for POJO Java projects — something that applies a configuration to a `LayerConfiguration.Builder` rather than existing as a separate parallel object.  And we should have a separate `WarLayerConfigurator` (or `WebAppLayerConfigurator`?) that takes an exploded war and applies a suitable configuration to a `LayerConfiguration.Builder`.  These different variants of `addFiles` could be put on a `LayerConfigurationBuilderHelper` helper class.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java [hunk 1]
```diff
@@ -60,6 +64,40 @@ String getName() {
...
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter only the files satisfying the filter will be added, unless the files are
+     *     directories
+     * @param basePathInContainer directory in the layer into which the source contents are added
+     * @param layerBuilder the {@link LayerConfiguration.Builder}
+     * @throws IOException error while listing directories
+     * @throws NotDirectoryException if {@code sourceRoot} is not a directory
+     */
+    @VisibleForTesting
+    static void addFilesRoot(
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java]: Looks like all of `JavaContainerBuilder`, `LayerConfiguration.Builder`, and `JavaLayerConfigurations.Builder` should be re-worked. I feel the current API is not very friendly for some of our use cases, where files in the same directory hierarchy should be separated out into different layers. Also, the lowest level `addEntry` and `addEntryRecursive` of `LayerConfiguration.Builder` is constrained in that it's very cumbersome to recursively copy files into the container root. (For that matter, you have to list files first and then call `addEntryRecursive` multiple times.)

[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java]: > I think how we could rework those APIs is to have `LayerConfiguration.Builder` only add the low-level sourceFile-extractionPath pairs,  and to have a higher-level helper

This makes sense. The low-level pair (with no recursion) is very limited but acceptable assuming the users won't work on the low-level `LayerConfiguration` directly. The helper can solve the recursive copy with filtering.

Now the question is, where this help should be placed by which name. I realized there is one annoyance with helper methods in another class: if such methods are `static`, they should accept a builder instance of  `LayerConfiguration.Builder` as an argument to work on it. This may be OK but feels a bit awkward personally, which makes me think maybe I should just put the helper methods under `LayerConfiguration.Builder` directly. If we don't have the helper methods as `static` (i.e., a stateful helper for building a `LayerConfiguration`), it becomes another `LayerConfiguration` builder that is more versatile than the low-level `LayerConfiguration.Builder`.

Also, currently `JavaLayerConfigurations` is another abstract layer on top of `LayerConfiguration`, which can also act as a helper, so we need to consider that too for deciding a suitable place for the helper class.

**Comment by chanseokoh** [comment]
Had a chat with @coollog. Some decisions made for now:

- Not changing `LayerConfigurations`
- Not replacing `JavaLayerConfigurations` with `JibContainerBuilder` yet
- Not using `JibContainerBuilder` in `MavenLayerConfiguration` or `GradleLayerConfiguration` yet

We can work on this PR now. Also,

- Remove dedicated methods for each layer from `JavaLayerConfigurations`. Instead, expose `LayerType` and have a single method for each operation. That is, no more `addClassFile()` and `addExtraFile()`, but have a single `addFile(LayerType.CLASSES, ...)`.
- Add to `JavaLayerConfigurations` a add helper that copies a directory to target recursively with filtering.

I realized we didn't talk about `JavaLayerConfigurationsHelper.fromExplodedWar`, but we can revisit that after the changes above if necessary.

Let me know if anything is missing or there's a concern.

**Comment by chanseokoh** [comment]
Still tying up some loose ends.

**Comment by chanseokoh** [comment]
OK, ready to review. See if this change makes sense.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java [hunk 2]
```diff
@@ -70,94 +73,57 @@ private Builder() {
...
+     * example, if {@code sourceRoot} is {@code /usr/home}, {@code /usr/home/passwd} exists locally,
+     * and {@code basePathInContainer} is {@code /etc}, then the image will have {@code
+     * /etc/passwd}.
      *
-     * @param sourceFile the source file to add to the layer
-     * @param pathInContainer the path in the container file system corresponding to the {@code
-     *     sourceFile}
+     * @param layerType the layer to add files into
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter only the files satisfying the filter will be added, unless the files are
```
Comment (TadCordle) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java]: Maybe reword to describe the parameter itself instead of what's affected by it, e.g. `filter that determines which files (not directories) should be added`

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 3]
```diff
@@ -75,87 +71,76 @@ static JavaLayerConfigurations getForProject(
...
-        logger.info("\t'" + classesOutputDirectory + "' (not found, skipped)");
-        continue;
-      }
-      logger.info("\t'" + classesOutputDirectory + "'");
-      try (Stream<Path> classFileStream = Files.list(classesOutputDirectory.toPath())) {
-        classFileStream.forEach(classesFiles::add);
-      }
-    }
-    if (classesFiles.isEmpty()) {
+    classesDirectories
```
Comment (TadCordle) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: This may be a small personal preference, but I found this logic (L90-101) easier to read when it was all under the same loop, rather than filtering out/iterating over missing and existing files separately.

Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: I see. I will revert it to do the logging in the same loop.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenLayerConfigurationsTest.java [hunk 4]
```diff
@@ -194,132 +201,6 @@ public void testGetForProject_nonDefaultAppRoot() throws IOException {
...
-        .add(sourceRoot.resolve("example/dir/C.class"), basePath.resolve("example/dir/C.class"));
-    Mockito.verify(fileToLayerAdder)
-        .add(sourceRoot.resolve("example/C.class"), basePath.resolve("example/C.class"));
-    Mockito.verify(fileToLayerAdder)
-        .add(sourceRoot.resolve("test/resources/leaf"), basePath.resolve("test/resources/leaf"));
-    Mockito.verify(fileToLayerAdder)
-        .add(sourceRoot.resolve("test/D.class"), basePath.resolve("test/D.class"));
-    Mockito.verifyNoMoreInteractions(fileToLayerAdder);
-  }
-
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenLayerConfigurationsTest.java]: That's true, but I think it's also meaningful to have the tests, since `JavaLayerConfigurationsHelperTest` only tests the `JavaLayerConfigurationsHelper.fromExplodedWar()` method, while here, they test `MavenLayerConfigurations.getForProject()` at a higher level, so they test things at different levels.
